Fixed identity matrix issue

ncave opened the pull request:
@dfarnham Great job porting this! Here is a minor fix to make it work with an identity matrix.

ncave committed:
Fixed identity matrix issue

dfarnham commented:
Thanks for pointing out the identity matrix problem and proposing a fix to avoid the current assertion of n > 0 in svd_dcopy().

The general solution of recognizing and assigning the singular values to the input diagonal for these matrix forms is the direction I’m leaning.

A runtime overflow exception can still occur with the current type of n:usize and subtracting when n==0, but there’s no data to copy in this case either, so this really needs to be detected upstream.

Thanks again and I hope you can workaround this for the time being!

Best,
Dave


> On Oct 15, 2022, at 2:07 PM, Слава Україні! Героям слава! ***@***.***> wrote:
> 
> @dfarnham <https://github.com/dfarnham> Great job porting this! Here is a minor fix to make it work with an identity matrix.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/dfarnham/svdlibrs/pull/1 <https://github.com/dfarnham/svdlibrs/pull/1>
> Commit Summary
> 
> f4c2a2b <https://github.com/dfarnham/svdlibrs/pull/1/commits/f4c2a2bc43b98dda62ff7f13d8fd8df4b185fc48> Fixed identity matrix issue
> File Changes (1 file <https://github.com/dfarnham/svdlibrs/pull/1/files>)
> M src/lib.rs <https://github.com/dfarnham/svdlibrs/pull/1/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759> (4)
> Patch Links:
> 
> https://github.com/dfarnham/svdlibrs/pull/1.patch <https://github.com/dfarnham/svdlibrs/pull/1.patch>
> https://github.com/dfarnham/svdlibrs/pull/1.diff <https://github.com/dfarnham/svdlibrs/pull/1.diff>
> —
> Reply to this email directly, view it on GitHub <https://github.com/dfarnham/svdlibrs/pull/1>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAOMDBH7PCQXTR7X6XSA74DWDMFJJANCNFSM6AAAAAARGBMGTE>.
> You are receiving this because you were mentioned.
>

ncave committed:
Added identity matrix test

ncave commented:
@dfarnham
Thanks, I've changed it to a conditional, and added a test. That's what the original `dcopy` does [when n = 0](https://github.com/lucasmaystre/svdlibc/blob/master/svdutil.c#L352).
I know it only produces one singular value (cause it probably stops early), but that's how it is, at least it doesn't throw now.

dfarnham commented:
@ncave Thanks for the improvements!
